File: Pharmaceutical_Company_Example/Behavioral/C++/Template_Method.cpp
#include <iostream>
#include <string>

// Abstract Class with Template Method
class RegulatoryProcess {
public:
  void execute_process(const std::string& drug_name) {
    std::cout << "Starting the regulatory process for drug: " << drug_name << std::endl;
    perform_evaluation(drug_name);
    perform_approval(drug_name);
    finalize_process(drug_name);
    std::cout << "Regulatory process completed for drug: " << drug_name << std::endl;
  }
  
  // Abstract Methods
  virtual void perform_evaluation(const std::string& drug_name) = 0;
  virtual void perform_approval(const std::string& drug_name) = 0;
  
  // Final Method
  void finalize_process(const std::string& drug_name) {
    std::cout << "Finalizing process for drug: " << drug_name << std::endl;
  }
  
  virtual ~RegulatoryProcess() = default;
};

// Concrete Class: Standard Approval Process
class StandardApprovalProcess : public RegulatoryProcess {
public:
  void perform_evaluation(const std::string& drug_name) override {
    std::cout << "Evaluating drug: " << drug_name << " using Standard Evaluation." << std::endl;
  }
  
  void perform_approval(const std::string& drug_name) override {
    std::cout << "Approving drug: " << drug_name << " using Standard Approval Process." << std::endl;
  }
};

// Concrete Class: Accelerated Approval Process
class AcceleratedApprovalProcess : public RegulatoryProcess {
public:
  void perform_evaluation(const std::string& drug_name) override {
    std::cout << "Evaluating drug: " << drug_name << " using Accelerated Evaluation." << std::endl;
  }
  
  void perform_approval(const std::string& drug_name) override {
    std::cout << "Approving drug: " << drug_name << " using Accelerated Approval Process." << std::endl;
  }
};

// Example Usage
int main() {
  // Create instances of the concrete classes
  StandardApprovalProcess standard_process;
  AcceleratedApprovalProcess accelerated_process;
  
  // Execute processes
  standard_process.execute_process("Test Small Molecule 01");
  accelerated_process.execute_process("Test Small Molecule 02");
  
  return 0;
}

File: Pharmaceutical_Company_Example/03_Behavioral/Java/Template_Method.java
// Define Abstract Class with Template Method
abstract class RegulatoryProcess {
    public final void executeProcess(String drugName) {
        System.out.println("Starting the regulatory process for drug: " + drugName);
        performEvaluation(drugName);
        performApproval(drugName);
        finalizeProcess(drugName);
        System.out.println("Regulatory process completed for drug: " + drugName);
    }

    // Abstract Methods (to be implemented by subclasses)
    protected abstract void performEvaluation(String drugName);
    protected abstract void performApproval(String drugName);

    // Concrete Method (can be overridden but has default behavior)
    protected void finalizeProcess(String drugName) {
        System.out.println("Finalizing process for drug: " + drugName);
    }
}

// Concrete Class: Standard Approval Process
class StandardApprovalProcess extends RegulatoryProcess {
    @Override
    protected void performEvaluation(String drugName) {
        System.out.println("Evaluating drug: " + drugName + " using Standard Evaluation.");
    }

    @Override
    protected void performApproval(String drugName) {
        System.out.println("Approving drug: " + drugName + " using Standard Approval Process.");
    }
}

// Concrete Class: Accelerated Approval Process
class AcceleratedApprovalProcess extends RegulatoryProcess {
    @Override
    protected void performEvaluation(String drugName) {
        System.out.println("Evaluating drug: " + drugName + " using Accelerated Evaluation.");
    }

    @Override
    protected void performApproval(String drugName) {
        System.out.println("Approving drug: " + drugName + " using Accelerated Approval Process.");
    }
}

// Main class to test the Template Method pattern
public class Main {
    public static void main(String[] args) {
        // Create instances of the concrete classes
        RegulatoryProcess standardProcess = new StandardApprovalProcess();
        RegulatoryProcess acceleratedProcess = new AcceleratedApprovalProcess();

        // Execute processes
        standardProcess.executeProcess("Test Small Molecule 01");
        System.out.println();
        acceleratedProcess.executeProcess("Test Small Molecule 02");
    }
}
